lib/Runtime/Types/TypePropertyCache.cpp

```diff
@@ -172,7 +172,9 @@ namespace Js
                 &propertyIndex,
                 &isInlineSlot,
                 &isMissing,
-                &prototypeObjectWithProperty))
+                &prototypeObjectWithProperty)
+            || (prototypeObjectWithProperty != nullptr
+                && prototypeObjectWithProperty != propertyObject))
```

Comment:
The TypePropertyCache can be populated in multiple occasions. In this repro case, calls to “super.x” go through these lines in GetProperty_Internal():

                PropertyValueInfo::Set(info, requestContext->GetLibrary()->GetMissingPropertyHolder(), 0);
                CacheOperators::CachePropertyRead(instance, requestContext->GetLibrary()->GetMissingPropertyHolder(), isRoot, propertyId, true, info, requestContext);

Now that the cache is populated with requestContext->GetLibrary()->GetMissingPropertyHolder() as “prototypeObjectWithProperty”, next call to “this.x” (having same “instance”) will be a cache hit (with prototypeObjectWithProperty != nullptr), whereas it should be a cache miss + update.


---
In reply to: [127714797](https://github.com/Microsoft/ChakraCore/pull/3336#discussion_r127714797) [](ancestors = 127714797)